File: output/cpp/tests/nested_braces.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


#include <iostream>
#include <string>
#include <cassert>
#include <map>
#include <vector>

// Test: Deeply nested braces and braces-in-strings should not confuse the body closer.

class NestedBracesFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    NestedBracesFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class NestedBracesFrameContext {
public:
    NestedBracesFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    NestedBracesFrameContext(NestedBracesFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class NestedBracesCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<NestedBracesFrameEvent> forward_event;
    std::unique_ptr<NestedBracesCompartment> parent_compartment;

    explicit NestedBracesCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<NestedBracesCompartment> clone() const {
        auto c = std::make_unique<NestedBracesCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class NestedBraces {
private:
    std::vector<std::unique_ptr<NestedBracesCompartment>> _state_stack;
    std::unique_ptr<NestedBracesCompartment> __compartment;
    std::unique_ptr<NestedBracesCompartment> __next_compartment;
    std::vector<NestedBracesFrameContext> _context_stack;

    void __kernel(NestedBracesFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            NestedBracesFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                NestedBracesFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    NestedBracesFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(NestedBracesFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Start") {
            _state_Start(__e);
        }
    }

    void __transition(std::unique_ptr<NestedBracesCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Start(NestedBracesFrameEvent& __e) {
        if (__e._message == "run_test") {
            // String containing braces
            std::string s1 = "string with { braces }";
            std::string s2 = "nested { { { braces } } }";
            std::string s3 = "{\"json\": \"in string\"}";

            // Multi-level nesting
            std::map<int, std::map<std::string, int>> result;
            for (int i = 0; i < 3; i++) {
                result[i] = {{"level", i}};
            }

            if (result.size() == 3) {
                printf("PASS: Nested braces handled correctly\n");
            } else {
                printf("FAIL: Wrong result count\n");
                assert(false);
            }
        }
    }

public:
    NestedBraces() {
        __compartment = std::make_unique<NestedBracesCompartment>("Start");
        NestedBracesFrameEvent __frame_event("$>");
        NestedBracesFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void run_test() {
        NestedBracesFrameEvent __e("run_test");
        NestedBracesFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }
};

int main() {
    NestedBraces s;
    s.run_test();
    return 0;
}

File: output/javascript/tests/nested_braces.mjs

// Test: Deeply nested braces and braces-in-strings should not confuse the body closer.

export class NestedBracesFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class NestedBracesFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class NestedBracesCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new NestedBracesCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class NestedBraces {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new NestedBracesCompartment("Start");
        this.__next_compartment = null;
        const __frame_event = new NestedBracesFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new NestedBracesFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new NestedBracesFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new NestedBracesFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    run_test() {
        const __e = new NestedBracesFrameEvent("run_test", null);
        const __ctx = new NestedBracesFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    _state_Start(__e) {
        if (__e._message === "run_test") {
            // Object with nested braces
            const d = {outer: {inner: {deep: 1}}};

            // String containing braces
            const s1 = "string with { braces }";
            const s2 = "nested { { { braces } } }";
            const s3 = '{"json": "in string"}';

            // Template literal with braces
            const name = "world";
            const s4 = `hello ${name} with { braces }`;
            const s5 = `${{key: "value"}}`;

            // Multi-level nesting
            const result = {};
            for (let i = 0; i < 3; i++) {
                result[String(i)] = {level: i, data: {items: [1, 2, 3]}};
            }

            if (Object.keys(result).length === 3) {
                console.log("PASS: Nested braces handled correctly");
            } else {
                console.log("FAIL: Wrong result count");
                throw new Error("test failed");
            }
        }
    }
}

const s = new NestedBraces();
s.run_test();
